---
title: "Summarize updated MT-DNA results (SNP- and Gene-based)"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
      toc: true
      toc_depth: 4
---

```{r, message=F, warning=F}
rm(list=ls())
library(PheWAS)
# install.packages(c("dplyr","tidyr","ggplot2","MASS","meta","ggrepel","DT"))
# devtools::install_github("PheWAS/PheWAS")

library(plotly)
library(forestplot)
library(geomtextpath)
library(colorspace)
library(scales)
library(grid)
library(khroma)
library("formatdown")
#library(readxl)
#library(filesstrings)
#library(UpSetR)

library(purrr)
library(data.table)
library(ComplexUpset)
library(knitr)
library(tidyr)
library(magrittr)
library(readr)
# library(xlsx)
library(readxl)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(stringi)
library(stringr)
```

```{r}
# define folder directory 
manu_supp_dir <- "/Users/jzhou/Library/CloudStorage/Box-Box/MVP/MT-MVP/manuscript/SupplementaryTablesFigures/"
manu_dir <- "/Users/jzhou/Library/CloudStorage/Box-Box/MVP/MT-MVP/manuscript/"
analysis_dir <- "/Users/jzhou/Library/CloudStorage/Box-Box/MVP/MT-MVP/analysis/updated/"
```

## Load Info
- Summary of phecode Ns in each population
```{r, message=F}
phecode_summ <- fread("../summary/MVP_summaries/summary_phecodes_by_GIA.sex.pop.vPAD.txt",
                      colClasses = c(phecode = "character")) %>%
  dplyr::rename(RACE = GIA, SEX = sex, POP = pop) %>%
  dplyr::select(-COHORT)
```

- Extract cardio-metabolic and liver phecodes for a summary table (supplemenatry table in the manuscript)
```{r}
#Phecodes related to cardiometabolic with >= 500 cases in any population
phecodes_analyzed <- fread("../summary/MVP_summaries/cardiometabolic_liver_phecodes.txt",
                           colClasses = c(phecode = "character")) %>%
  left_join(phecode_summ) %>%
  group_by(phecode) %>%
  filter(SEX == "allsex" & 
         POP == "all" & 
         RACE %in% c("AFR_GIA", "AMR_GIA", "EUR_GIA", "EAS_GIA")) %>%
  mutate(RACE = gsub("_GIA", "", RACE)) %>%
  filter(any(n_cases >= 500)) %>%
  mutate(`N (%)` = paste0(n_cases, " (", round(prop_cases*100, 2), ")")) %>%
  mutate(`N (%)` = ifelse(n_cases < 500, "< 500", `N (%)`)) %>%
  dplyr::select(-n_cases, -prop_cases) %>%
  pivot_wider(names_from = "RACE", values_from = `N (%)`, names_prefix = "N (%) ") %>%
  #322 actual from 365 relevant
  dplyr::select(phecode, description, group, `N (%) EUR`, `N (%) AFR`, `N (%) AMR`, `N (%) EAS`) 

table(phecodes_analyzed$group)
phecodes_analyzed %>% filter(group %in% c("congenital anomalies", "digestive"))

# write_xlsx(phecodes_analyzed %>% data.frame(.), 
# "../summary/MVP_summaries/phecodes_analyzed.xlsx", row.names=F)
```
- Multiallelic SNPs to exclude
```{r}
multiallelic_snps <- c("AX-37116935", "AX-83589457", 
                       "AX-98077966", "AX-98077957", 
                       "AX-83452320", "AX-98077978", 
                       "AX-98078001")
```

- Load rsIDs, gene names, and other annotations from hmtvar database
```{r}
anno <- read_xlsx("../hmtvar/Suppl.xlsx", skip = 1) %>%
  dplyr::select(ID, BP, dbSNP, Locus, Pathogenicity, 
                MITOMAP_disease = `MITOMAP disease`) %>%
  left_join(read_xlsx("../summary/aa_changes.xlsx"))

anno_mvp <-  read_xlsx("../../../MVP mitochondria SNPs annotation 20200710.xlsx")  %>% 
  select(`dbSNP RS ID`, `NC_012920 (rCRS) Physical Position`) %>%
  rename(BP = `NC_012920 (rCRS) Physical Position`) 

anno <-  anno %>% 
         left_join(anno_mvp, by = c("BP")) %>%
         select(-dbSNP) %>% 
         rename(dbSNP = `dbSNP RS ID`)
```

- Load MT base info for solar plots
```{r}
(refflat <- read_table(
  paste(analysis_dir, "summary/gene_definitions/gencode_mt_refFlat.txt", sep = ""),
  col_names = F) %>% 
  select(1, 5, 6, 9) %>%
  set_names(c("gene", "start", "end", "ctrl")) %>% 
  mutate(gene = gsub("MT-", "", gene)))

pos_gene <- tibble(
    POS = seq(0, 16567), 
    ctrl = 1) %>%
  full_join(refflat) %>%
  filter(POS >= start & POS <= end) %>%
  dplyr::select(-ctrl, start, end) %>% 
  group_by(POS) %>% 
  dplyr::slice(1) %>%
  mutate(bp = POS) %>%
  ungroup()

(refflat_length <- refflat %>%
  dplyr::rename(Gene = gene) %>%
  mutate(gene_length = end - start,
         add_pos = cumsum(gene_length) - 71) %>%
  mutate(small_gene = ifelse(gene_length < 350, 1, 0)) %>%
  mutate(pos.rt = 360 -((start + gene_length/2)/(16023))*360))

(refflat_length_small_plothelp <- refflat_length %>%
  filter(small_gene == 1) %>%
  mutate(plot_pos = case_when(
    Gene %in% c("TF", "TM", "TC", "TD", "ATP8", "TL2") ~ start + gene_length/2 + 60,
    Gene %in% c("TI", "TA", "TH", "TT")  ~ start + gene_length/2 - 60,
    Gene %in% c("TY")  ~ start + gene_length/2 + 120,
    Gene %in% c("TW")  ~ start + gene_length/2 - 120,
    TRUE ~ start + gene_length/2)
  ))

(bprange <- refflat %>%
  filter(!grepl("^T", gene)) %>%
  dplyr::select(-ctrl) %>%
  mutate(BP = paste0( start, "-", end)) %>%
  mutate(SetID = paste0("MT-", gene)) %>% 
  dplyr::select(-gene, -start, -end))
```
## Define data encoding keys and color schema for plots
- Define data encoding keys
```{r}
biomarker_key <- c("a1c" = "HbA1c",
                   "tg" = "Triglycerides",
                   "bmi" = "BMI",
                   "ldlc" = "LDL",
                   "hdlc" = "HDL",
                  
                   "tg_adj" = "Triglycerides",
                   "ldlc_adj" = "LDL",
                   "hdlc_adj" = "HDL",
                   
                   "tg_adj_INT" = "Triglycerides",
                   "ldlc_adj_INT" = "LDL",
                   "hdlc_adj_INT" = "HDL")

race_key <- c("AFR_GIA" = "African",
             "EUR_GIA" = "European",
             "EAS_GIA" = "East Asian", 
             "AMR_GIA" = "Admixed American",
             "TE_META" = "Trans-Ethnic Meta-analysis")

sex_key <- c("males" = "Males",
            "females" = "Females",
            "allsex" = "Males + Females")

sex_key2 <- c("males" = "Males",
            "females" = "Females",
            "allsex" = "")

pop_key <- c("t2d" = "T2D",
             "nondiab" = "Non DM",
             "all" = "All")

sex_pop_key <- c("t2dmales" = "Males with T2D",
                 "t2dfemales" = "Females with T2D",
                 "t2dallsex" = "All T2D",
                 "nondiabmales" = "Non-diabetic Males",
                 "nondiabfemales" = "Non-diabetic Females",
                 "nondiaballsex" = "All Non-diabetics",
                 "allmales" = "All Males",
                 "allfemales" = "All Females",
                 "allallsex"= "All")

sex_pop_key2 <- c("t2dmales" = "Males with T2D",
                 "t2dfemales" = "Females with T2D",
                 "t2dallsex" = "T2D",
                 "nondiabmales" = "Non-diabetic Males",
                 "nondiabfemales" = "Non-diabetic Females",
                 "nondiaballsex" = "Non-diabetics",
                 "allmales" = "Males",
                 "allfemales" = "Females",
                 "allallsex"= "All")
```

- Functions and colors to use in plots
```{r}
### some code is from stephen turner solarplot.R, modified ####
addgenelabel <- function(bp) {
  gene <- left_join(data.frame(bp), pos_gene) 
  gene$gene[is.na(gene$gene) & gene$bp < 577] <- "Control-Region"
  gene <- gene %>% dplyr::select(gene)
  return(gene$gene)
}

lighten_col <- function(col){
  col1 = RGB(col2rgb(col)[1]/255, col2rgb(col)[2]/255, col2rgb(col)[3]/255)
  col2 = RGB(1,1,1)
  newcol <- colorspace::mixcolor(0.4, col1, col2)
  newcol_char <- rgb(newcol@coords[,1], newcol@coords[,2], newcol@coords[,3])
  return(newcol_char)
}

 # Set colors for each gene
colours_solid <- c("Control-Region" = "lightblue4", 
                   "tRNA" = "magenta4",
                   "TF" = "magenta4", 
                   "RNR1" = "mediumaquamarine", 
                   "TV" = "magenta4",
                   # "Non-Coding" = "sienna4", 
                   "RNR2" = "aquamarine",
                   "TL1" = "magenta4",
                   "ND1" = "magenta", 
                   "TI" = "magenta4",
                   "TQ" = "magenta4",
                   "TM" = "magenta4",
                   "ND2" = "mediumblue", 
                   "TW" = "magenta4",
                   "TA" = "magenta4",
                   "TN" = "magenta4",
                   "TC" = "magenta4",
                   "TY" = "magenta4",
                   "CO1" = "olivedrab", 
                   "TS1" = "magenta4",
                   "TD" = "magenta4",
                   "TK" = "magenta4",
                   "CO2" = "orange2", 
                   "ATP8" = "orchid4", 
                   "ATP6" = "red3", 
                   "CO3" = "royalblue2", 
                   "TG" = "magenta4",
                   "ND3" = "palegreen4",
                   "TR" = "magenta4",
                   "ND4L" = "grey0", 
                   "ND4" = "pink4", 
                   "TH" = "magenta4",
                   "TS2" = "magenta4",
                   "TL2" = "magenta4",
                   "ND5" = "yellow4", 
                   "ND6" = "steelblue4",
                   "TE" = "magenta4",
                   "CYB" = "tan",
                   "TT" = "magenta4",
                   "TP" = "magenta4")
#colours <- t_col(colours_solid, name = names(colours_solid))
colours <- lapply(colours_solid, lighten_col) %>% unlist()
colours
show_col(unique(colours))
```
## Summarize mtSNVs association analysis results 
###  SNP-based results (Phewas and biomarkers)
- Read in all analysis results, including ones with errors (i.e., analysis didn't finish successfully)
- Remove multi-alleletic SNPs   
- Keep analysis results in GIA ancestry groups 
- Keep analysis results adjusted for 15PCs 
- Keep analysis results among most recent cohort "vPDA"
```{r}
d_snp_all <- fread(
  paste(analysis_dir, 
        "SNPbased/ancestries/MTSNVs/results.SNPbased.firth.GIA.15PC.vPAD.txt", sep = "")) %>% 
  # remove old unadjusted biomarkers
  filter(!(trait %in% c("sbp", "dbp", "hdlc", "ldlc", "tg"))) %>% 
  full_join(fread("../sNPbased/ancestries/MTSNVs/results.SNPbased.biomarker_adjustments.txt") %>%
            # add adjusted biomarkers
            filter(grepl("bp_adj$|adj_INT", trait)) %>% 
            dplyr::select(-`#CHROM`, -TEST, 
                          -T_STAT, -filename, 
                          -ASSOC_TYPE, -TEST_TYPE, 
                          -ANCESTRY, -DESCRIPTOR)) %>% 
  filter(COHORT == "vPAD" & NPC == "15PC"  & grepl("GIA", RACE)) %>% 
  # remove multi-alleletic snps
  filter(!(ID %in% multiallelic_snps)) %>%
  dplyr::select(-COHORT, -NPC) %>%
  mutate(phecode = gsub("_", ".", gsub("phe_", "", trait))) 
```

- Join phecode info and add snp annotation
```{r}
d_snp_all <- d_snp_all %>% 
  left_join(phecode_summ, by = c("RACE", "SEX", "POP", "phecode")) %>%
  left_join(anno %>% dplyr::rename(POS = BP))
```

## SNP-level ancestry-specific results

### Further filtering results 
- More analyses were done than needed, so 
  * Keep only first level subgroups 
  * A1c analysis only makes sense in DM stratified groups
- Filter analysis didn't converge and with no P

- Keep analysis results for phecodes with > 500 cases
- SNP-level analysis is for MAF > 0.001
- Note that FDR p-value needs to be calculated after filtering 
  * Multiple testing correction applied separately for each ancestry/sex/population analysis
- Calculate FDR adjusted p-value
   * overall 4 ancestries, 
   * 4 ancestries + SEX subgroups, 
   * 4 ancestries + diabetes subgroups 

```{r, message=F}
d_snp <- d_snp_all %>% 
  # filter analysis with no error code
  filter(ERRCODE == ".") %>%
  dplyr::select(-ERRCODE) %>%
  # filter analysis didn't converge with no P
  filter(!is.na(P)) %>%
  distinct() %>%
  # only include 1-level subgroups (e.g., males, t2d, not males w/ t2d)
  filter(SEX == "allsex" | POP == "all") %>% 
  # a1c needs to be stratified by dm only
  filter(!(POP == "all" & trait %in% c("a1c", "g65_hb", "g75_hb")))  %>%
  # keep analysis results for phecodes with > 500 cases
  filter(n_cases >= 500 | TRAIT_TYPE == "other_traits") %>%
  # SNP-level analysis is for MAF > 0.001
  filter(ALT_FREQS >= 0.001 & ALT_FREQS <= 0.999) %>%
  ungroup() %>%
  group_by(RACE, SEX, POP) %>%
  mutate(p_adjust_fdr = p.adjust(P, "BH")) 
```


### Table 2: 9 associations  
- Significant SNP-level results by controlling FDR < 0.05
```{r}
d_sig_fdr <- d_snp %>%
  filter(p_adjust_fdr < 0.05)   
  

# #Use EUR allsex all (the most tests are performed in this group) to estimate most conservative FDR pval threshold
# #Maybe also estimate the least conservative and have a dashed rectangle?
# p_sig_fdr <- d_sig_fdr %>%
#   filter(POP == "all" & SEX == "allsex" & RACE == "EUR_GIA") %>%
#   mutate(dis = abs(0.05-p_adjust_fdr)) %>%
#   arrange(dis) %>%
#   slice(1) %>%
#   ##filter(dis <= 0.01) %>%
#   dplyr::select(RACE, SEX, POP, p_adjust_fdr, P, dis) %>%
#   distinct() %>%
#   .$P
# p_sig_fdr
#   
#   
# d_sig_fdr <- d_sig_fdr %>%
#   filter(p_adjust_fdr < 0.05) 

```

```{r}
d_sig_fdr_eur <- d_sig_fdr %>% 
  filter(POP == "all" & SEX == "allsex") %>%
  # keep only 244.4 (leaves) not 244
  filter(phecode != "244") %>%
  filter(RACE == "EUR_GIA")

d_sig_fdr_afr <- d_sig_fdr %>% 
  filter(POP == "all" & SEX == "allsex") %>%
  # keep only 244.4 (leaves) not 244
  filter(phecode != "244") %>%
  filter(RACE == "AFR_GIA")

snp_table1_eur <- (d_snp_all %>% 
                     left_join(d_snp %>% select(POS, RACE, phecode, SEX, POP, p_adjust_fdr), 
                     by = c("POS", "SEX", "RACE",  "phecode",  "POP"))) %>% 
  filter(POP == "all" & SEX == "allsex") %>%
  select(-POP, -SEX) %>%
  right_join(d_sig_fdr_eur %>% ungroup %>% select(POS, phecode), by = c("POS", "phecode")) %>%
  left_join(phecodes_analyzed %>% select(phecode, description), by = "phecode") %>%
  select(RACE, POS, 
         dbSNP, REF, 
         ALT, A1, OBS_CT,
         Locus, ALT_FREQS, 
         OR, `LOG(OR)_SE`, 
         P, p_adjust_fdr,
         n_cases, prop_cases, 
         variant_type, aa_change,
         phecode, description,) %>%
  mutate(RACE = factor(str_replace(RACE, "_GIA", ""), levels = c("EUR", "AFR", "AMR", "EAS")),
         altallele = A1,
         refallele = ifelse(REF == A1, ALT, REF),
         `BP/REF/ALT` = str_c(POS, "/", refallele, "/", altallele, sep = ""),
         MAF = ifelse(ALT_FREQS > 0.5, 1-ALT_FREQS, ALT_FREQS),
         Annotation = ifelse(aa_change == "NA", variant_type, 
                             str_c(variant_type, ", ", aa_change, sep = ""))) %>%
  arrange(POS, phecode, RACE)  %>%
  select(-aa_change, -variant_type, -POS, -REF, -ALT) %>%
  mutate(`# Cases / N (%)` = ifelse(n_cases < 500, 
                              " < 500", 
                              paste(formatC(n_cases, format = "d", big.mark = ","), "/",
                                    formatC(OBS_CT, format = "d", big.mark = ","),
                                    paste0("(", round(prop_cases*100, 1), "%)"), sep=" "))) %>%
  mutate( LCI = signif(exp(log(OR) - 1.96*`LOG(OR)_SE`), 2), 
          UCI = signif(exp(log(OR) + 1.96*`LOG(OR)_SE`), 2),
          P = scientific(P, digits = 3), 
          p_adjust_fdr = round(p_adjust_fdr, digits = 2),
          `OR (95% CI), P-value` = 
            ifelse(ALT_FREQS < 0.001 | ALT_FREQS > 0.999 | n_cases < 500,
                   "-", paste0(round(OR, 2), " (", LCI, "-" , UCI, "), ", P)),
          `FDR Adjusted P-value` = ifelse(`OR (95% CI), P-value` == "-", "-", p_adjust_fdr)) %>%
  select(-LCI, -UCI, -P, -OR, -`LOG(OR)_SE`, -n_cases, -prop_cases) %>%
  # mutate(`Alt Freq` = paste0(round(ifelse(ALT_FREQS > 0.5, 1-ALT_FREQS, ALT_FREQS)*100, 2), "%")) %>%
  mutate(`Alt AF` = ifelse(ALT_FREQS < 0.001 | ALT_FREQS > 0.999, 
                             "< 0.1%", 
                           paste0(round(MAF*100, 2), "%"))) %>%
  select(-ALT_FREQS) %>%
  mutate(`Description, Phecode` = paste0(description, ", ", phecode)) %>%
  select(-description, -phecode) %>%
  select(dbSNP, `BP/REF/ALT`, 
         Locus, Annotation, 
         `Description, Phecode`, 
         RACE, `# Cases / N (%)`, 
         `Alt AF`, `OR (95% CI), P-value`,
         `FDR Adjusted P-value`) %>%
  ungroup() %>%
  group_by(dbSNP) %>% 
  #Don't repeat  
  mutate(dbSNP = ifelse(row_number() == 1, dbSNP, ""),
         # `BP/REF/ALT` = ifelse(row_number() == 1, `BP/REF/ALT`, ""),
         Locus = ifelse(row_number() == 1, Locus, ""),
         Annotation = ifelse(row_number() == 1, Annotation, ""),
         `Description, Phecode` = ifelse(row_number() == 1, `Description, Phecode`, "")) 

snp_table1_afr <- (d_snp_all %>% 
                     left_join(d_snp %>% select(POS, RACE, phecode, SEX, POP, p_adjust_fdr), 
                     by = c("POS", "SEX", "RACE",  "phecode",  "POP")))  %>% 
  filter(POP == "all" & SEX == "allsex") %>%
  select(-POP, -SEX) %>%
  right_join(d_sig_fdr_afr %>% ungroup %>% select(POS, phecode), by = c("POS", "phecode")) %>%
  left_join(phecodes_analyzed %>% select(phecode, description), by = "phecode") %>%
  select(RACE, POS, 
         dbSNP, REF, 
         ALT, A1, OBS_CT,
         Locus, ALT_FREQS, 
         OR, `LOG(OR)_SE`, 
         P, p_adjust_fdr,
         n_cases, prop_cases, 
         variant_type, aa_change,
         phecode, description) %>%
  mutate(RACE = factor(str_replace(RACE, "_GIA", ""), levels = c("EUR", "AFR", "AMR", "EAS")),
         altallele = A1,
         refallele = ifelse(REF == A1, ALT, REF),
         `BP/REF/ALT` = str_c(POS, "/", refallele, "/", altallele, sep = ""),
         MAF = ifelse(ALT_FREQS > 0.5, 1-ALT_FREQS, ALT_FREQS),
         Annotation = ifelse(aa_change == "NA", variant_type, 
                             str_c(variant_type, ", ", aa_change, sep = ""))) %>%
  arrange(POS, phecode, RACE)  %>%
  select(-aa_change, -variant_type, -POS, -REF, -ALT) %>%
  mutate(`# Cases / N (%)` = ifelse(n_cases < 500, 
                              " < 500", 
                              paste(formatC(n_cases, format = "d", big.mark = ","), "/",
                                    formatC(OBS_CT, format = "d", big.mark = ","),
                                    paste0("(", round(prop_cases*100, 1), "%)"), sep=" "))) %>%
  mutate( LCI = signif(exp(log(OR) - 1.96*`LOG(OR)_SE`), 2), 
          UCI = signif(exp(log(OR) + 1.96*`LOG(OR)_SE`), 2),
          P = scientific(P, digits = 3), 
          p_adjust_fdr = round(p_adjust_fdr, digits = 2),
          `OR (95% CI), P-value` = ifelse(ALT_FREQS < 0.001 | ALT_FREQS > 0.999 | n_cases < 500,
                                          "-", paste0(round(OR, 2), " (", LCI, "-" , UCI, "), ", P)),
          `FDR Adjusted P-value` = ifelse(`OR (95% CI), P-value` == "-", "-", p_adjust_fdr)) %>%
  select(-LCI, -UCI, -P, -OR, -`LOG(OR)_SE`, -n_cases, -prop_cases) %>%
  # mutate(`Alt Freq` = paste0(round(ifelse(ALT_FREQS > 0.5, 1-ALT_FREQS, ALT_FREQS)*100, 2), "%")) %>%
  mutate(`Alt AF` = ifelse(ALT_FREQS < 0.001 | ALT_FREQS > 0.999, 
                             "< 0.1%", paste0(round(MAF*100, 2), "%"))) %>%
  select(-ALT_FREQS) %>%
  mutate(`Description, Phecode` = paste0(description, ", ", phecode)) %>%
  select(-description, -phecode) %>%
  select(dbSNP, `BP/REF/ALT`, 
         Locus, Annotation, 
         `Description, Phecode`, 
         RACE, `# Cases / N (%)`, 
         `Alt AF`, `OR (95% CI), P-value`,
         `FDR Adjusted P-value`) %>%
  ungroup() %>%
  group_by(dbSNP) %>% 
  #Don't repeat  
  mutate(dbSNP = ifelse(row_number() == 1, dbSNP, ""),
         # `BP/REF/ALT` = ifelse(row_number() == 1, `BP/REF/ALT`, ""),
         Locus = ifelse(row_number() == 1, Locus, ""),
         Annotation = ifelse(row_number() == 1, Annotation, ""),
         `Description, Phecode` = ifelse(row_number() == 1, `Description, Phecode`, "")) 

snp_table1 <- bind_rows(snp_table1_eur, snp_table1_afr)
# write_csv(snp_table1, "../tables/table1.csv")
```

```{r}
# output all report mtSNVs
mtsnvs <- d_sig_fdr %>% 
  filter(POP != "nondiab") %>% 
  ungroup %>% 
  select(POS, RACE, SEX, POP) %>% 
  distinct() %>%
  arrange(POS) %>% 
  mutate(POS = paste("MT", POS, sep = "")) %>% 
  mutate(RACE = str_replace(RACE, "_GIA", "")) %>% 
  mutate(population = case_when(SEX == "allsex" & POP == "all" ~ "All", 
                                SEX == "allsex" & POP == "t2d" ~ "T2D",
                                SEX == "males" & POP == "all" ~ "Males",
                                SEX == "females" & POP == "all" ~ "Females")) %>% 
  select(-SEX, -POP)

# write_csv(mtsnvs, "../tables/reported_mtsnvs.csv")
```

```{r}
# output all report mtSNVs
unique_phecodes <- d_sig_fdr %>% 
  filter(POP != "nondiab") %>% 
  ungroup %>% 
  select(phecode) %>% 
  distinct() %>%
  left_join(phecodes_analyzed %>% select(phecode, description, group)) %>%
  arrange(phecode) 

write_csv(unique_phecodes, "../tables/unique_phecodes.csv")
```

- %cases distribution (not included as figure in the manuscript)
```{r, eval = F}
# Use position = position_dodge() 
bar_phecode <- d_snp %>% 
  filter(POP == "all" & SEX == "allsex") %>%
  select(-POP, -SEX) %>%
  right_join(d_sig_fdr %>% ungroup %>% select(POS, phecode), by = c("POS", "phecode")) %>%
  left_join(phecodes_analyzed %>% select(phecode, description), by = "phecode") %>%
  arrange(POS, phecode, RACE, prop_cases) %>%
  select(RACE, 
         n_cases, prop_cases, 
         phecode, description) %>%
  distinct()

bar_phecode <- bar_phecode %>% 
  bind_rows(tibble(RACE = "EAS_GIA", n_cases = 0,
                   prop_cases = 0, phecode = "259.3", 
                   description = "Delay in sexual development and puberty NEC")) %>%
  bind_rows(tibble(RACE = "EAS_GIA", n_cases = 0,
                   prop_cases = 0, phecode = "253.4", 
                   description = "Anterior pituitary disorders")) %>%
  arrange(phecode, RACE)

library(viridis)
library(hrbrthemes)
p <- ggplot2::ggplot(bar_phecode %>% mutate(prop_cases = prop_cases*100), 
                     aes(x = RACE, y = prop_cases, fill = RACE)) +
  geom_bar(position  = position_dodge(0.8), width = 0.7, stat = "identity") +
  facet_wrap(~phecode, scales = "free") +
    theme_ipsum() +
    theme(legend.position="none") +
    xlab("")
p
```

- MAF distribution (not included as figure in the manuscript)
```{r, eval = F}
# Use position = position_dodge() 
bar_maf <- d_snp %>% 
  filter(POP == "all" & SEX == "allsex") %>%
  select(-POP, -SEX) %>%
  right_join(d_sig_fdr %>% ungroup %>% select(POS, phecode), by = c("POS", "phecode")) %>%
  left_join(phecodes_analyzed %>% select(phecode, description), by = "phecode") %>%
  arrange(POS, phecode, RACE, prop_cases) %>%
  select(RACE, POS, dbSNP, ALT_FREQS)  %>%
  mutate(ALT_FREQS = ifelse(ALT_FREQS > 0.5, (1-ALT_FREQS)*100, ALT_FREQS*100))

library(viridis)
library(hrbrthemes)
p <- ggplot2::ggplot(bar_maf, 
                     aes(x = RACE, y = ALT_FREQS, fill = RACE)) +
  geom_bar(position  = position_dodge(0.8), width = 0.7, stat = "identity") +
  facet_wrap(~POS, scales = "free") +
    theme_ipsum() +
    theme(legend.position="none") +
    xlab("")
p

```



## SNP-based interaction results (by T2D and Sex)
```{r}
#(Sex)
intx_sex <- read_xlsx("../SNPbased/interaction_sex_t2d/intx_result_sex_v2_glm.xlsx", 
                      col_types = c(rep("guess", 13), "text", rep("guess", 4)))
#(t2d)
intx_t2d <- read_xlsx("../SNPbased/interaction_sex_t2d/intx_result_t2d_v2_glm.xlsx", 
                      col_types = c(rep("guess", 13), "text", rep("guess", 4)))
```


```{r, include=F}
# msng <- full_join(intx_sex %>% mutate(intx_type = "SEX") %>% dplyr::select(-POP), 
#                   intx_t2d %>% mutate(intx_type = "POP") %>% dplyr::select(-SEX)) %>%
#   full_join(d_sig_fdr %>% mutate(table2=T)) %>%
#   filter(RACE != "TE_META") %>%
#   group_by(intx_type, RACE, ID, trait) %>%
#   mutate(n=n()) %>%
#   filter(n == 1 & table2 == T & is.na(intx_type)) %>%
#   filter(SEX != "allsex" | POP != "all")
# msng <- msng[,colSums(is.na(msng)) < nrow(msng)] %>%
#   left_join(pheinfo)
# #Why extra?
# allintx <- full_join(intx_sex %>% mutate(intx_type = "SEX") %>% dplyr::select(-POP), 
#                   intx_t2d %>% mutate(intx_type = "POP") %>% dplyr::select(-SEX)) %>%
#   distinct(RACE, intx_type, ID, trait, phecode, description)
```


### Figure 2: Forestplot in the manuscript 
- all interaction tests whether significant or not 
```{r}
#Add all three subgroups
intx_sex_broad <- intx_sex %>%
  dplyr::select(RACE, trait, ID, intx_sex_pval, POP, description) %>%
  left_join(d_snp) %>%
  mutate(subgrp = sex_key[SEX], studypop = pop_key[POP])

# Data for forestplot
dat_forestplot <- intx_sex_broad %>%
  select(-POP, 
         -studypop, 
         -Pathogenicity, 
         -MITOMAP_disease,  
         -TRAIT_TYPE, 
         -SE, 
         -SEX) %>% 
  right_join(d_snp_all %>% 
               filter(POP == "all") %>% 
               filter(paste(POS, phecode, RACE) %in% 
                      paste(intx_sex_broad$POS, intx_sex_broad$phecode, intx_sex_broad$RACE)) %>%
               select(-POP,
                      -Pathogenicity, 
                      -MITOMAP_disease,  
                      -TRAIT_TYPE,  
                      -SE, 
                      -SEX)) %>%
  mutate(intx_pval = coalesce(intx_sex_pval),
         BETA = coalesce(BETA, log(OR)),
         LCI = signif(exp(BETA - 1.96*`LOG(OR)_SE`), 2), 
         UCI = signif(exp(BETA + 1.96*`LOG(OR)_SE`), 2),
         P = ifelse(n_cases < 500, "-", signif(P, 2)),
         ORCI = paste0(signif(OR,2), " (", LCI, "-" , UCI, "), ", P),
         `Adj. P-value` = round(p_adjust_fdr, 3), 
         altallele = A1,
         refallele = ifelse(REF == A1, ALT, REF),
         `BP/REF/ALT` = str_c(POS, "/", refallele, "/", altallele, sep = ""),
         subgrp = ifelse(subgrp == "All", "Males + Females", subgrp),
         `description, phecode` = paste0("\n", description, ",\n", phecode),
         MAF = ifelse(ALT_FREQS > 0.05, 1-ALT_FREQS, ALT_FREQS),
         `ALT AF` = paste0(round(MAF*100, 2), "%"),
         Annotation = ifelse(aa_change == "NA", variant_type, 
                             str_c(variant_type, ", ", aa_change, sep = "")),
         intx_pval = scientific(intx_pval, digits = 3),
         intx_pval_print = ifelse(n_cases < 500, "-", as.character(intx_pval))) %>%
  group_by(trait, ID) %>% 
  #Don't repeat the interaction pval
  mutate(pval_reduced = 
           ifelse(any(intx_pval_print == "-"), "-", intx_pval_print)) %>%
  mutate(pval_reduced = ifelse(row_number() == 1, pval_reduced, "")) %>%
  # Don't repeat dbSNP on each row
  mutate(dbsnp_reduced = 
           ifelse(row_number() == 1, dbSNP, "")) %>% 
  # Don't repeat gene on each row
  mutate(gene_reduced = 
           ifelse(row_number() == 1, Locus, "")) %>% 
  # Don't repeat pos/alt/ref on each row
  mutate(locus_reduced = 
           ifelse(row_number() == 1, `BP/REF/ALT`, "")) %>% 
  # Don't repeat the description on each row
  mutate(description_reduced = 
           ifelse(row_number() == 1, `description, phecode`, "")) %>% 
  # Don't repeat the %AL Freq on each row
  mutate(alt_freq_reduced = 
           ifelse(row_number() == 1, `ALT AF`, "")) %>%
  # Don't repeat the %AL Freq on each row
  mutate(annotation_reduced = 
           ifelse(row_number() == 1, Annotation, "")) %>% 
  # Don't repeat the race/subgroup
  mutate(RACE = str_replace(RACE, "_GIA", "")) %>%
  mutate(RACE = factor(RACE, levels = c("EUR", "AFR"))) %>%
  mutate(race_reduced = ifelse(row_number() == 1, as.character(RACE), "")) 

dat_forestplot_eur <- dat_forestplot %>%
  filter(RACE == "EUR") %>%
  mutate(subgrp = ifelse(is.na(subgrp), "Females", subgrp)) %>%
  mutate(subgrp = factor(subgrp, levels = c("Males + Females", "Females", "Males"))) %>%
  ungroup() %>%
  arrange(RACE, POS, trait)
  
dat_forestplot_afr <- dat_forestplot %>%
  filter(RACE == "AFR") %>% 
  mutate(subgrp = ifelse(is.na(subgrp), "Females", subgrp)) %>%
  mutate(subgrp = factor(subgrp, levels = c("Males + Females", "Males", "Females"))) %>%
  ungroup() %>%
  arrange(RACE, POS, trait, subgrp)
  
dat_forestplot <- bind_rows(dat_forestplot_eur, dat_forestplot_afr)

from_rmeta <- 
  structure(list(
    mean  = c(NA, dat_forestplot$OR), 
    lower = c(NA, dat_forestplot$LCI),
    upper = c(NA, dat_forestplot$UCI)),
    .Names = c("mean", "lower", "upper"), 
    row.names = c(NA, -34L), 
    class = "data.frame")

dat_forestplot <- dat_forestplot %>% 
  mutate(case_label = ifelse(n_cases < 500, 
                              " < 500", 
                              paste(formatC(n_cases, format = "d", big.mark = ","), "/",
                                    formatC(OBS_CT, format = "d", big.mark = ","),
                                    paste0("(", round(prop_cases*100, 1), "%)"), sep=" "))) 
  
tabletext <- cbind(
  c("Ancestry", as.character(dat_forestplot$race_reduced)),
  c("BP/REF/ALT", dat_forestplot$locus_reduced),
  c("dbSNP", dat_forestplot$dbsnp_reduced),
  c("Gene", dat_forestplot$gene_reduced),
  c("Annotation", dat_forestplot$annotation_reduced),
  c("Description, Phecode", as.character(dat_forestplot$description_reduced)),
  c("Population", as.character(dat_forestplot$subgrp)),
  c("# Cases / N (%)", dat_forestplot$case_label),
  c("Alt AF", as.character(dat_forestplot$`ALT AF`)),
  c("OR (95% CI), P-value", as.character(dat_forestplot$ORCI)),
  c("Adj. P-value", as.character(dat_forestplot$`Adj. P-value`)),
  c("Int. P-value", as.character(dat_forestplot$pval_reduced)))
```

```{r}
hlines <- list(NULL, 
               gpar(col="#444444", lwd = 4), 
               NULL, 
               NULL,
               # gpar(col="#444444", columns=2:8, lty="dashed"), NULL, NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444", lwd = 4), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444", lwd = 4))

fp1 <- forestplot(tabletext, 
           from_rmeta,
           is.summary= c(T, rep(F, nrow(tabletext)-1)),
           align = c(rep("l", ncol(tabletext)-2), "r", "r"),
           # is.summary=c(T,F,F, T,F,F, T,F,F, T,F,F) #,
           # xticks = c(0.5,0.75,1,1.25,1.50,2.0,3.0,5, 8),
           # graphwidth = unit(50,'mm'),
           xlog = T, 
           new_page = T,
           # boxsize = 0.2,
           graph.pos = 8,
           boxsize = 0.3,
           lwd.ci = 3,
           colgap = unit(.05, "npc"),
           txt_gp = fpTxtGp(cex = 2, 
                            ticks = gpar(cex = 1.8), 
                            xlab = gpar(cex = 1.8),
                            label = list(gpar(fontface = "bold"))),
           col = fpColors(box="royalblue",line="darkblue"),
           #hrzl_lines = T,
           ###hrzl_lines = gpar(col="#444444"),
           hrzl_lines = hlines,
           # lineheight = unit(1, "cm"), 
           # lineheight = 'lines',
           xlab = "OR",
           xticks.digits = 1,
           # xticks = c(0.025, 0.125, 0.5, 1, 2,4,8, 16)
           line.margin = 0.3,
           # ###title=c("Interaction Tests with Sex or T2D Status")
           ) #+ #theme(strip.placement = NULL) #fixes error but plot becomes empty
  #theme_bw()
png(paste("../images/Forest_nosubgrp_inclnonsig_simpler.png"),
 width = 45,
 height = 25,
 units = 'in',
 res = 600)
fp1
dev.off()

```

### Figure 3: Forestplot  in the manuscript (T2D vs non-diabetes)
- all interaction tests whether significant or not 
```{r}
#Add all three subgroups
intx_t2d_broad <- intx_t2d %>%
  dplyr::select(RACE, trait, ID, intx_t2d_pval, SEX, description) %>%
  left_join(d_snp) %>%
  mutate(subgrp = pop_key[POP], studypop=sex_key[SEX]) %>%
  group_by(trait, POS) %>% 
  mutate(subgrp_min = subgrp[P == min(P)]) 

# Data for forestplot
dat_forestplot <- intx_t2d_broad %>%
  select(-POP, 
         -Pathogenicity, 
         -MITOMAP_disease,  
         -TRAIT_TYPE,  
         -SE, 
         -SEX) %>% 
  right_join(d_snp_all %>% 
               filter(SEX == "allsex") %>% 
               filter(paste(POS, phecode, RACE) %in% 
                      paste(intx_t2d_broad$POS, intx_t2d_broad$phecode, intx_t2d_broad$RACE)) %>%
               select(-POP,
                      -Pathogenicity, 
                      -MITOMAP_disease,  
                      -TRAIT_TYPE,  
                      -SE, 
                      -SEX)) %>%
  filter(phecode != "276.4") %>%
  group_by(trait, ID) %>% 
  mutate(subgrp_min = subgrp_min[row_number()==1]) %>% 
  ungroup() %>%
  mutate(intx_pval = coalesce(intx_t2d_pval),
         RACE = str_replace(RACE, "_GIA", ""),
         BETA = coalesce(BETA, log(OR)),
         LCI = signif(exp(BETA - 1.96*`LOG(OR)_SE`), 2), 
         UCI = signif(exp(BETA + 1.96*`LOG(OR)_SE`), 2),
         P = ifelse(n_cases < 500, "-", signif(P, 2)),
         ORCI = paste0(signif(OR,2), " (", LCI, "-" , UCI, "), ", P),
         `Adj. P-value` = round(p_adjust_fdr, 3), 
         IDname = paste0(dbSNP, " (", Locus, ")"),
         altallele = A1,
         refallele = ifelse(REF == A1, ALT, REF),
         `BP/REF/ALT` = str_c(POS, "/", refallele, "/", altallele, sep = ""),
         description = ifelse(phecode == "440.21", 
                   "\n Atherosclerosis of native arteries of \n the extremities with ulceration or gangrene", 
                              description),
         `description, phecode` = paste0("\n", description, ",\n", phecode),
         MAF = ifelse(ALT_FREQS > 0.5, 1-ALT_FREQS, ALT_FREQS),
         `ALT AF` = paste0(round(MAF*100, 2), "%"),
         Annotation = ifelse(aa_change == "NA", variant_type, 
                             str_c(variant_type, ", ", aa_change, sep = "")),
         intx_pval_print = ifelse(n_cases < 500, "-", as.character(signif(intx_pval, 2)))) %>%
  mutate(RACE = factor(RACE, levels = c("EUR", "AFR", "AMR"))) %>%
  mutate(subgrp = ifelse(is.na(subgrp), "Non DM", subgrp),
    subgrp = factor(coalesce(subgrp), levels = c("All", "T2D", "Non DM")),
    subgrp_min = factor(subgrp_min, levels = c("All", "T2D", "Non DM"))) %>% 
  group_by(trait, POS) %>% 
  #Don't repeat the interaction pval
  mutate(pval_reduced = 
           ifelse(any(intx_pval_print == "-"), "-", intx_pval_print)) %>%
  mutate(pval_reduced = ifelse(row_number() == 1, pval_reduced, "")) %>%
  # Don't repeat dbSNP on each row
  mutate(dbsnp_reduced = 
           ifelse(row_number() == 1, dbSNP, "")) %>% 
  # Don't repeat gene on each row
  mutate(gene_reduced = 
           ifelse(row_number() == 1, Locus, "")) %>% 
  # Don't repeat pos/alt/ref on each row
  mutate(locus_reduced = 
           ifelse(row_number() == 1, `BP/REF/ALT`, "")) %>% 
  # Don't repeat the description on each row
  mutate(description_reduced = 
           ifelse(row_number() == 1, `description, phecode`, "")) %>% 
  # Don't repeat the %AL Freq on each row
  mutate(alt_freq_reduced = 
           ifelse(row_number() == 1, `ALT AF`, "")) %>%
  # Don't repeat the %AL Freq on each row
  mutate(annotation_reduced = 
           ifelse(row_number() == 1, Annotation, "")) %>% 
  mutate(race_reduced = 
           ifelse(row_number() == 1, as.character(RACE), "")) %>% 
  ungroup() %>%
  arrange(subgrp_min, RACE, POS, trait, subgrp)  %>% 
  filter(subgrp_min != "Non DM")

from_rmeta <- 
  structure(list(
    mean  = c(NA, dat_forestplot$OR), 
    lower = c(NA, dat_forestplot$LCI),
    upper = c(NA, dat_forestplot$UCI)),
    .Names = c("mean", "lower", "upper"), 
    row.names = c(NA, -25L), 
    class = "data.frame")

dat_forestplot <- dat_forestplot %>% 
  mutate(case_label = ifelse(n_cases < 500, 
                              " < 500", 
                              paste(formatC(n_cases, format = "d", big.mark = ","), "/",
                                    formatC(OBS_CT, format = "d", big.mark = ","),
                                    paste0("(", round(prop_cases*100, 1), "%)"), sep=" "))) 

tabletext <- cbind(
  c("Ancestry", as.character(dat_forestplot$race_reduced)),
  c("BP/REF/ALT", dat_forestplot$locus_reduced),
  c("dbSNP", dat_forestplot$dbsnp_reduced),
  c("Gene", dat_forestplot$gene_reduced),
  c("Annotation", dat_forestplot$annotation_reduced),
  c("Description, Phecode", as.character(dat_forestplot$description_reduced)),
  c("Population", as.character(dat_forestplot$subgrp)),
  c("# Cases / N (%)", dat_forestplot$case_label),
  c("Alt AF", as.character(dat_forestplot$`ALT AF`)),
  c("OR (95% CI), P-value", as.character(dat_forestplot$ORCI)),
  c("Adj. P-value", as.character(dat_forestplot$`Adj. P-value`)),
  c("Int. P-value", as.character(dat_forestplot$pval_reduced)))
```

```{r}
hlines <- list(
               NULL,
               gpar(col="#444444", lwd = 4), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444"), 
               NULL, 
               NULL,
               gpar(col="#444444", lwd = 4))

fp1 <- forestplot(tabletext, 
           from_rmeta,
           is.summary = c(T, rep(F, nrow(tabletext)-1)),
           align = c(rep("l", ncol(tabletext)-2), "r", "r"),
           # is.summary=c(T,F,F, T,F,F, T,F,F, T,F,F) #,
           # xticks = c(0.5,0.75,1,1.25,1.50,2.0,3.0,5, 8),
           # graphwidth = unit(50,'mm'),
           xlog = T, 
           new_page = T,
           # boxsize = 0.2,
           graph.pos = 8,
           boxsize = 0.3,
           lwd.ci = 3,
           colgap = unit(.05, "npc"),
           txt_gp = fpTxtGp(cex = 2, 
                            ticks = gpar(cex = 1.8), 
                            xlab = gpar(cex = 1.8),
                            label = list(gpar(fontface = "bold"))),
           col = fpColors(box="royalblue",line="darkblue"),
           #hrzl_lines = T,
           ###hrzl_lines = gpar(col="#444444"),
           hrzl_lines = hlines,
           # lineheight = unit(1, "cm"), 
           # lineheight = 'lines',
           xlab = "OR",
           xticks.digits = 1,
           # xticks = c(0.025, 0.125, 0.5, 1, 2,4,8, 16)
           line.margin = 0.3
           ) 
  #theme_bw()

png(paste("../images/Forest2_nosubgrp_inclnonsig_simpler.png"),
    width = 45,
    height = 20,
    units = 'in',
    res = 400)
fp1 
dev.off()
```



## Gene-based results (phecodes and biomarkers)
Read in and clean results
```{r, message=F}
d_gene <- fread("../genebased/MT/SKATO.GIA.15PC.all_traits.result.txt") %>% 
  filter(COHORT == "vPAD" & grepl("GIA", RACE)) %>% 
  filter(!(trait %in% c("sbp", "dbp", "hdlc", "ldlc", "tg", "bmi", "a1c", "g65_hb", "g75_hb"))) %>% #remove old
  full_join(fread("../genebased/MT/SKATO.other_traits.result.addbiomarker_adjustments.raw.txt") %>%
              filter(grepl("bp_adj$|adj_INT|a1c|bmi|g65|g75", trait))) %>% #add biomarkers
  distinct() %>%
  filter(N.Marker.Test > 1 & observations >= 500 & (n_cases >= 500 | is.na(n_cases)) & !is.na(P.value)) %>%
  dplyr::select(-N.Marker.All, -COHORT, -DESCRIPTOR) %>%
  mutate(phecode = gsub("_", ".", gsub("phe_", "", trait))) %>%
  left_join(pheinfo %>% dplyr::select(phecode, description, group)) %>%
  filter(SEX == "allsex" | POP == "all") %>%
  filter(!(POP == "all" & trait %in% c("a1c", "g65_hb", "g75_hb"))) #a1c needs to be stratified by dm only
length(unique(d_gene$SetID))
```

Significant gene-based results (FDR)
```{r}
fdr = 0.05
gene_sig_fdr <- d_gene %>%
  
  group_by(RACE, SEX, POP) %>%
  mutate(P.value.adj.fdr = p.adjust(P.value, "BH")) %>%
  filter(P.value.adj.fdr <= fdr)
```

Write out traits/genes for MTgene replication
```{r}
MTgene_forUKBrepl <- gene_sig_fdr %>%
  filter(RACE == "EUR_GIA") %>%
  distinct(SetID, P.value, SEX, POP, N.Marker.Test, observations, trait, 
           prop_cases, n_cases,  phecode, description, group)
#fwrite(MTgene_forUKBrepl, "../replication_input_data/MTgene_forUKBreplication.txt", row.names=F, sep = "\t")
```

Write out MT-gene significant results suitable for table 4.
```{r}
names(gene_sig_fdr)
table4 <- gene_sig_fdr %>%
  left_join(bprange) %>%
  mutate(Symbol = gsub("MT-", "", SetID)) %>%
  #left_join(mymap) %>%
  #left_join(mitocarta_mt_anno %>% 
  #            mutate(anno_match = 1, 
  #                  Symbol = gsub("COX", "CO", Symbol),
  #                  Symbol = gsub("CYT", "CY", Symbol))) %>%
  mutate(Ancestry = gsub("_GIA", "", RACE),
         Gene = gsub("MT-", "", SetID),
         bpstart = as.numeric(gsub("-.*", "", BP)),
         Sex = sex_key[SEX],
         Population = pop_key[POP],
         Population_sex = ifelse(Population == "All", Sex, Population),
         `% Cases` = round(prop_cases*100, 1), 
          ##Trait = coalesce(phecode, trait),
         Trait = coalesce(biomarker_key[trait], phecode),
         P = signif(P.value, 3),
         P.value.adj.fdr = round(P.value.adj.fdr, 3)) %>%
  left_join(pheinfo) %>%
  ungroup() %>%
  mutate(Ancestry = factor(Ancestry, 
                           levels = c("EUR", "AFR", "AMR", "EAS"), 
                           labels = c("EUR", "AFR", "AMR", "EAS"))) %>%
  dplyr::select(Population_sex, Ancestry, BP, Gene, N.Marker.Test, # MCARTA2.0_score, 
         # Primary_Subcellular_Localization_2015 =  HPA_PrimarySubcellularLocalization_2015,
         # Gene_Description = Description,
         `% Cases`, Trait, description, Chapter = group, bpstart, P, P.value.adj.fdr) %>%
    arrange(Population_sex, Ancestry, bpstart, P) %>%
  dplyr::select(-bpstart) %>% 
  # remove association with 272, as 272.1 was also reported
  filter(Trait != "272")
  
fwrite(table4, "../tables/table4_mt_gene.csv", row.names=F)
```

Combined gene-based solar plot (gray genes version)
Data setup
```{r}
### MT Gene-based data #####

dat <- d_gene %>% filter(SEX == "allsex" | POP == "all") # %>% dplyr::select(-bonf)
dat_sig <- gene_sig_fdr #%>% dplyr::select(-bonf)

# Creates and stores negative log p as a new variable
dat$neglogp <- -log10(dat$P.value)

# Adds line to y axis
#dat$extraline <- -log10(5E-5)
#extraline <- -log10(0.05/nrow(dat))

nphecodes <- length(unique(dat$phecode))

dat_plot <- dat %>%
  left_join(dat_sig) %>%
  mutate(sig = ifelse(!is.na(P.value.adj.fdr), 1, 0)) %>%
  dplyr::rename(Gene = SetID) %>%
  mutate(Gene = gsub("MT-", "", Gene)) %>%
  left_join(refflat_length) %>%
  arrange(Gene,as.numeric(phecode)) %>%
  #mutate(runlength = cumsum(unique(gene_length)) - gene_length[1]) %>%
  group_by(Gene) %>%
  mutate(length_nrow_ratio = dplyr::n()/gene_length,
       phe_pos = row_number()/length_nrow_ratio,
       POS = phe_pos + start)
  #filter(phecode %in% dat_sig$phecode) %>%
  #filter(P.value != 0 & !is.na(P.value))
#table(dat_plot$phecode)
```

Colours
```{r}
###colours_ct <- colours[!grepl("^T", names(colours))]
colours_gw <- colours
colours_gw[1:length(colours_gw)] <- rep(c("gray50", "gray60", "gray70"), 100)[1:length(colours)]
##show_col(colours_ct)
##colours_pop <- c("gray20", "#E16666", "#6666E1", "#98B998", "#FFBA66")

# Define colours for population key
colours_pop <- c("gray20", "#E16666", "#6666E1", "#98B998", "#B9ADA3") #lighten peachpuff4
names(colours_pop) <- sort(sex_pop_key2[grepl("All", sex_pop_key)])

#colours_group <- lapply(c("purple1", "purple4","deeppink2", "deeppink4"), lighten_col) %>% unlist()
##colours_group <- lapply(c("purple1", "purple4","deeppink2", "deeppink4", "aquamarine4"), lighten_col) %>% unlist()

#Define colours for phecode group key
colours_group <- lapply(c("purple1", "aquamarine4", "darkorange3", "purple4", "deeppink2"), lighten_col) %>% unlist()
names(colours_group) <- c("circulatory system", "endocrine/metabolic", "digestive", "congenital anomalies", "biomarkers")
show_col(colours_group)
#colorscale_plot <- c(colours_gw, colours_pop)

# Define colourscale including both population and phecode group (they use the same aesthetic mapping)
colorscale_plot <- c(colours_group, colours_pop)
```

Plot
```{r}

c_radius <- 6.5
dat_plot2 <- dat_plot %>%
  mutate(Population = sex_pop_key2[paste0(POP,SEX)],
         `Genetically Inferred Ancestry` = race_key[RACE]) %>% 
  mutate(
    #Gene_orig = Gene, Gene = ifelse(grepl("^T", Gene), "tRNA", Gene),
       gene_lbl = ifelse(sig, Gene, NA),
       description = ifelse(is.na(description), biomarker_key[trait], description),
       description_shortened = gsub("[;,]* and ", "/", description),
       #phecode_title1 = paste0(phecode, "\n(", description_shortened, ")"), #two lines label
       phecode_title2 = paste0(phecode, "-", description_shortened),   #one line label
       #lbl1 = ifelse(sig, phecode_title1, ""),
       lbl2 = ifelse(sig, phecode_title2, ""),
       lbl2 = ifelse(is.na(group) & sig, biomarker_key[trait], lbl2),
       lbl3 = ifelse(sig & ((Gene %in% c("ATP6", "ATP8", "CO3", "ND6", "ND1") & phecode != "426") |
                              (Gene == "RNR1" & trait == "tg")),
                     paste0(lbl2, "\n(", gene_lbl, ")"), lbl2),  #Add the gene label for significant points that are close to the gene border
       lbl_extra = ifelse(sig, paste0("(", gene_lbl, ", ", Population, ")"), NA),
       lbl4 = ifelse(sig, paste0(lbl2, "\n", lbl_extra), NA), #Add the gene AND population label for all points.
       #psize = ifelse(sig, 3, 1),
       points_gray = ifelse(sig, NA, POS), #NONsignificant points
       points_colored = ifelse(sig, POS, NA), #significant points
       group = ifelse(is.na(group), "biomarkers", group)) #%>%
  # ungroup() %>%
  # # mutate(degree = 360*(POS/max(POS)),
  # #        ang = degree - 270 - ifelse(degree <= 180, 180, 0)) %>%
  # group_by(Gene) %>%
  # #mutate(gene_ymax = max(neglogp)) %>%
  # ungroup()
  # #arrange(desc(neglogp)) %>%
  # #hist(dat_plot2$degree)

fplot <- ggplot(dat_plot2) +
  ylim(0, max(dat_plot$neglogp) + c_radius + 1) +
  coord_polar(direction = -1, clip = "off") +
  #guides(fill=guide_legend(ncol=2)) +
  guides(fill = "none") +
  guides(size = "none") +
  #guides(color=guide_legend(ncol=2)) +
  #guides(color="none", size = "none") +
  #geom_line(aes(x, 5,color = "red"), data = lines, size = 1.5) +
  #facet_grid(.~phecode_title) +
  theme_minimal() +
  #geom_hline(aes(yintercept=extraline + c_radius), color = "red") +
  geom_hline(aes(yintercept=c_radius + 1), color = "gray", size = 0.1) +
  geom_hline(aes(yintercept=c_radius + 2), color = "gray", size = 0.1) +
  geom_hline(aes(yintercept=c_radius + 3), color = "gray", size = 0.1) +
  geom_hline(aes(yintercept=c_radius + 4), color = "gray", size = 0.1) +
  geom_hline(aes(yintercept=c_radius + 5), color = "gray", size = 0.1) +
  geom_hline(aes(yintercept=c_radius + 6), color = "gray", size = 0.1) +
  geom_hline(aes(yintercept=c_radius + 7), color = "gray", size = 0.1) +
  geom_hline(aes(yintercept=c_radius + 8), color = "gray", size = 0.1) +
  geom_hline(aes(yintercept=c_radius + 9), color = "gray", size = 0.1) +
  geom_hline(aes(yintercept=c_radius + 10), color = "gray", size = 0.1) +
  geom_rect(data=refflat, # Gene bands
            aes(xmin=start, xmax=end, ymin=c_radius-1.2, ymax=c_radius-0.2, fill=gene),
            inherit.aes = F, show.legend = F) +
  geom_textpath(data=refflat_length, # Gene labels on the band (for larger genes)
                aes(x=start + gene_length/2, y=c_radius - 0.7,
                    label = ifelse(small_gene, NA, Gene),
                    #label = Gene,
                    angle=ifelse(small_gene, 90, 0),
                    size = ifelse(small_gene, 1.5, 5)),
                inherit.aes=F, show.legend=F) +
  geom_point(aes(x=points_gray, y=neglogp + c_radius, shape = `Genetically Inferred Ancestry`, #plot non-significant points (by phecode group)
                 color=group), size=1) + #, color = "gray65") +
  geom_point(aes(x=points_colored, y=neglogp + c_radius, color = group, #Population, #plot significant points in different color scale (population)
                 shape = `Genetically Inferred Ancestry`, size=psize), size=3.5 ) +
  geom_text_repel(aes(x=POS, y=neglogp + c_radius, label = lbl4), #lbl3), #label significant points
                show.legend=F, max.overlaps=Inf, max.time = 20, color = "gray1", size=3,
                #min.segment.length = .1, force = 3, nudge_y=0.2,
                box.padding = unit(0.75, "lines")) +
  # geom_text_repel(data=refflat_length, #label small genes - ggrepel won't work for this due to polar coordinates
  #               aes(x=start + gene_length/2, y=c_radius - 0.2,
  #                   angle = 90 - ifelse(pos.rt >= 180, pos.rt - 180, pos.rt),
  #                   label = ifelse(small_gene, Gene, "")),
  #                   #angle=ifelse(small_gene, 90, 0),
  #               #size = ifelse(small_gene, 1.5, 5)),
  #               inherit.aes=F,
  #               max.overlaps = Inf,
  #               size=3, min.segment.length = 0,
  #               force_pull = 3,
  #               #ylim = c(c_radius+ 7.5, c_radius + 8.5), #NOT RESPECTED IN POLAR COORDINATES
  #               direction = "x", #nudge_y=0.5,
  #               show.legend=F) +
  geom_text(data=refflat_length_small_plothelp, #label small genes that don't fit well in the bands - define above
                aes(x=plot_pos, y=c_radius + .25,
                    angle = 90 - ifelse(pos.rt >= 180, pos.rt - 180, pos.rt),
                    label = Gene),
                    #angle=ifelse(small_gene, 90, 0),
                #size = ifelse(small_gene, 1.5, 5)),
                inherit.aes=F, size=3, show.legend=F) +
  scale_colour_manual(##values = c("black", "red", "blue", "green", "orange"),
                      ##values = c("gray20", "#E16666", "#6666E1", "#98B998", "#FFBA66"),
                      ######values = colorscale_plot, #includes both phecode group and population
                      values = colours_group,
                      #breaks = names(colorscale_plot)[grepl("All", names(colorscale_plot))],
                      ###"Population Phecode Type",
                      "Phecode Type",
                      #breaks = names(colours_ct),
                      #labels = colours_ct
                      ) +
  scale_fill_manual(values = colours_gw, #fill in the gene bands with gray colors
                      "",
                      breaks = names(colours_gw),
                      #labels = names(colours_gw)
                    ) +
  #xlab("Mitochondrial Base-Pair Location") +
  xlab("") +
  ylab("-log(p-value)") +
  #ggtitle("Negative Log P-value of Mitochondrial Hits") +
  ggtitle("Gene-based analysis") +
   theme(axis.text.x = element_blank(),
         axis.text.y = element_blank(),
         axis.ticks.y = element_blank(),
  #       legend.position = "none",
         panel.grid = element_blank()) +
  #theme(legend.key.size = unit(.7, 'cm')) +
  #scale_shape_manual(values = letters[univ$Id],
   #          guide = guide_legend(override.aes = list(alpha = 1,
    #                                                  size = 3))) +
  scale_shape_manual(values=c(15,16,17,9)) + #
  scale_size(range=c(2, 3.5)) # finicky
#fplot

# ggsave("../images/genebased_images/combined_fdr_solarplot_genebased_gray_group_v2.png", fplot,
#      w=12, h=10, dpi=400, bg = "white")
ggsave("../images/genebased_images/combined_fdr_solarplot_genebased_labelPop.png", fplot,
     w=12, h=10, dpi=400, bg = "white")

```

Summary of Number of markers per set
```{r}
gene_summary <- fread("../genebased/MT/SKATO.GIA.15PC.all_traits.result.txt") %>% 
  filter(COHORT == "vPAD") %>%
  filter(grepl("GIA", RACE)) %>% 
  filter(SEX == "allsex" & POP == "all") %>%
  filter(grepl("phe", trait)) %>%
  distinct(RACE, SetID, N.Marker.Test) %>%
  dplyr::rename(GIA = RACE) %>%
  mutate(GIA = gsub("_GIA", "", GIA),
         SetID = gsub("MT-", "", SetID)) %>%
  pivot_wider(id_cols = GIA, names_from = SetID, values_from = "N.Marker.Test") %>%
  data.frame()

##write.xlsx2(gene_summary, "../summary/MVP_summaries/N.Markers.summary.xlsx", row.names = F)
```

```{r}
# Initialize variables
m <- 137       # Genes IN endo/meta term
n <- 185      # Genes NOT IN endo/meta term
k <- 28       # Gene hits, that is sig mtSNVs
x <- c(0:28)  # Genes both IN endo/meta termterm and sig 'hits'

t <- 22 # number of hits in endo/meta term

# Use the dhyper built-in function for hypergeometric density
probabilities <- dhyper(x, m, n, k, log = FALSE)
probabilities
pvalue <- sum(probabilities[(t+1):length(x)])
pvalue
```
